src/command_modules/azure-cli-sql/azure/cli/command_modules/sql/custom.py

```diff
@@ -2310,3 +2315,119 @@ def virtual_cluster_list(
     return client.list()
+
+###############################################
+#              sql instance failover group    #
+###############################################
+
+
+# pylint: disable=too-few-public-methods
+class InstanceFailoverPolicyType(Enum):
+    automatic = 'Automatic'
+    manual = 'Manual'
+
+
+def instance_failover_group_create(
+        cmd,
+        client,
+        resource_group_name,
+        managed_instance,
+        instance_failover_group_name,
+        partner_managed_instance_name,
+        partner_resource_group,
+        failover_policy=InstanceFailoverPolicyType.automatic.value,
+        grace_period=1):
+    '''
+    Creates a failover group.
+    '''
+
+    managed_instance_client = get_sql_managed_instances_operations(cmd.cli_ctx, None)
+    # pylint: disable=no-member
+    primary_server = managed_instance_client.get(
+        managed_instance_name=managed_instance,
+        resource_group_name=resource_group_name)
+
+    partner_server = managed_instance_client.get(
+        managed_instance_name=partner_managed_instance_name,
+        resource_group_name=partner_resource_group)
+
+    # Build the partner server id
+    managed_server_info_pair = ManagedInstancePairInfo(
+        primary_managed_instance_id=primary_server.id,
+        partner_managed_instance_id=partner_server.id)
+    partner_region_info = PartnerRegionInfo(location=partner_server.location)
+
+    # Convert grace period from hours to minutes
+    grace_period = int(grace_period) * 60
+
+    if failover_policy == InstanceFailoverPolicyType.manual.value:
+        grace_period = None
+
+    return client.create_or_update(
+        resource_group_name=resource_group_name,
+        location_name=primary_server.location,
+        failover_group_name=instance_failover_group_name,
+        parameters=InstanceFailoverGroup(
+            managed_instance_pairs=[managed_server_info_pair],
+            partner_regions=[partner_region_info],
+            read_write_endpoint=InstanceFailoverGroupReadWriteEndpoint(
+                failover_policy=failover_policy,
+                failover_with_data_loss_grace_period_minutes=grace_period),
+            read_only_endpoint=InstanceFailoverGroupReadOnlyEndpoint(
+                failover_policy="Disabled")))
+
+
+def instance_failover_group_update(
+        instance,
+        failover_policy=None,
+        grace_period=None,):
+    '''
+    Updates the failover group.
+    '''
+
+    if failover_policy is not None:
+        instance.read_write_endpoint.failover_policy = failover_policy
+
+    if instance.read_write_endpoint.failover_policy == FailoverPolicyType.manual.value:
+        grace_period = None
+        instance.read_write_endpoint.failover_with_data_loss_grace_period_minutes = grace_period
+
+    if grace_period is not None:
+        grace_period = int(grace_period) * 60
+        instance.read_write_endpoint.failover_with_data_loss_grace_period_minutes = grace_period
```

Comment:
This is duplicated with failover_group_update. Please extract common logic, do NOT copy paste.